coordinator/src/services/translation/request/payload/factory.ts

```diff
@@ -0,0 +1,203 @@
+  getMessageIdFromBundle,
+  getMessageIdFromClaim,
+  getMessageIdFromTask,
+  identifyMessageType
+} from "../../common"
+import {convertCancellation} from "../cancel/cancellation"
+import {convertDispenseClaim} from "../dispense/dispense-claim"
+import {convertDispenseNotification} from "../dispense/dispense-notification"
+import {translateReleaseRequest} from "../dispense/release"
+import {convertParentPrescription} from "../prescribe/parent-prescription"
+import {convertTaskToDispenseProposalReturn} from "../return/return"
+import {createSendMessagePayload} from "../send-message-payload"
+import {convertTaskToEtpWithdraw} from "../withdraw/withdraw"
+
+type BundleTranslationResult = hl7V3.ParentPrescriptionRoot
+  | hl7V3.CancellationRequestRoot
+  | hl7V3.DispenseNotificationRoot
+  | hl7V3.DispenseClaimRoot
+
+type ParametersTranslationResult = hl7V3.NominatedPrescriptionReleaseRequestWrapper
+  | hl7V3.PatientPrescriptionReleaseRequestWrapper
+
+type TaskTranslationResult = hl7V3.DispenseProposalReturnRoot | hl7V3.EtpWithdrawRoot
+
+type ClaimTranslationResult = hl7V3.DispenseClaimRoot
+
+export type PayloadContent = BundleTranslationResult
+  | TaskTranslationResult
+  | ParametersTranslationResult
+  | ClaimTranslationResult
+
+type Payload<T extends PayloadContent> = {
+  content: T,
+  interactionId: hl7V3.Hl7InteractionIdentifier
+}
+
+type FactoryInput = fhir.Bundle | fhir.Task | fhir.Parameters | fhir.Claim
+
+export abstract class PayloadFactory {
+  protected abstract create(
+    fhirResource: FactoryInput,
+    logger?: pino.Logger
+  ): Payload<PayloadContent>
+
+  protected abstract getPayloadId(fhirResource: FactoryInput): string
+
+  static forBundle(): PayloadFactory {
+    return new BundleTranslationResultFactory()
+  }
+
+  static forTask(): PayloadFactory {
+    return new TaskTranslationResultFactory()
+  }
+
+  static forParameters(): PayloadFactory {
+    return new ParametersTranslationResultFactory()
+  }
+
+  static forClaim(): PayloadFactory {
+    return new ClaimTranslationResultFactory()
+  }
+
+  public makeSendMessagePayload(
+    fhirResource: FactoryInput,
+    headers: Hapi.Util.Dictionary<string>,
+    logger: pino.Logger
+  ): hl7V3.SendMessagePayload<PayloadContent> {
+    this.logIdentifiers(fhirResource, logger)
+
+    const messageId = this.getPayloadId(fhirResource)
+    const payload = this.create(fhirResource, logger)
+
+    return createSendMessagePayload(messageId, payload.interactionId, headers, payload.content)
+  }
+
+  /**
+   * Log resource identifier, to facilitate logs tracing on Splunk.
+   */
+  private logIdentifiers(fhirResource: FactoryInput, logger: pino.Logger) {
+    const logObject = {
+      payloadId: this.getPayloadId(fhirResource)
+    }
+    logger.info(logObject, "Creating Spine payload from FHIR resource")
+  }
+}
+
+class BundleTranslationResultFactory extends PayloadFactory {
+  getPayloadId(bundle: fhir.Bundle): string {
+    return getMessageIdFromBundle(bundle)
```

Comment:
is there a conceptual difference between `payloadId` and `messageId`? if not, can we sync up the names?

Reply:
Good spot-- `payloadId` is the identifier for the payload, while `messageId`s are generated by Apigee. I'll update those functions